demo/src/features/skybox/extract.rs

```diff
@@ -0,0 +1,51 @@
+rafx::declare_render_feature_extract_job!();
+
+use super::StaticResources;
```

Comment:
Drawing the skybox as a feature made several dozen lines into several hundred lines. I have a hard time agreeing with a change that does the same thing but is ~10x the amount of code. I wouldn't do it for example/teaching purposes either because I would not recommend someone to do it this way.

We could add a new resource that the old code fetches to drive whether the skybox is enabled and what material to draw. This would avoid anything skybox-specific being part of the interface for the graph building. In general this is how I would pass arbitrary configuration from user's high-level game logic into the render graph. (We can also use this method for configuring cameras/viewports.)

Reply:
BTW the "plugin" system is very 1.0, minimum viable product. The function calls that it has are simply what was needed to decouple the Renderer from the demo code and push it up into a crate. There may be better ways to provide the same hooks. (like adding resources that contain data.)

Reply:
The feature approach has more lines, yes. The actual _work_ (static resources, extract, prepare, draw) is more or less the same, it's just that the trait implementations of the plugin & jobs adds the overhead. However, there are 3 mitigating factors that I think are relevant:

1. Once you've seen one feature's boilerplate, you've seen them all. 

It's not going to be difficult for someone to zero in on the skybox code -- static resources are defined in `mod.rs`, the retrieval from the asset manager is in `extract`, dynamic descriptor sets are in `prepare`, and the submit calls are in `write`. 

The previous code required one to understand that the skybox materials are defined in `DemoStaticResources`,  they're retrieved from the asset manager in `render_graph_generator/mod.rs`, and the call to prepare descriptor sets & submit happens simultaneously in the middle of `opaque_pass.rs` between `OpaqueRenderPhase` and `TransparentRenderPhase` and is delegated to  a file `skybox.rs` in `features` but is not actually a feature.

2. There is a strong benefit in my mind towards reducing the number of different ways to configure a thing.

Consider the skybox. Implemented as a feature, it behaves the same as anything else. You can create a `RenderView` that selectively masks for it -- like a main player view (skybox enabled), a rear-view mirror in a car (skybox enabled), or a minimap (skybox disabled).

If it's not implemented as a feature, but is instead a resource of some type, we then need code to handle different resources per view. So we can put that onto the `RenderView` (e.g. `skybox_enabled`) and the render graph can iterate over all of the views and check if they have the `skybox_enabled` flag set, but at some point it's just _another_ way to describe "this is a thing that may or may not be drawn for a particular view". We'll still end up with dozens of more lines of code, but instead of it following the same pattern as all of the other features commands, it'll be different.

3. I think a large part of your aversion to making the skybox is a feature is that it's "just" a triangle.
 
I can easily imagine a higher fidelity skybox with e.g. multiple layers (to simulate parallax of things like clouds or stars), or an animated texture instead of static, or more complicated shader effects. In my mind, the render graph is specifically for compositing passes & screen-space post-processing effects together in a way that can efficiently manage GPU resources, whereas features are for what is drawn inside of passes or as a per-view post-processing effect. For example, in this slide from the Destiny talk, the render graph feels like the equivalent of Destiny's _"render passes"_ -- shading, tone mapping & resolve, etc. The skybox is a _"render stage"_ -- it's _"execute submission for a specific view for a specific stage of the frame pipeline where we might have content-driven data"_ 

![image](https://user-images.githubusercontent.com/4888530/114225779-96efbc00-9927-11eb-8018-6c92bf235ce1.png)

Reply:
Been thinking about this today:
 * I still think that the 10x increase in amount of code is not worth the consistency gained. (Also that it's "negative teaching" because this isn't the way I would recommend someone to add a full-screen triangle pass.)
 * A skybox seems to match the green text rather than the yellow text. I doubt they went through their nodes system for their "atmosphere" pass.
 * The current form of the skybox is ideally drawn after `OpaqueRenderPhase` and before `TransparentRenderPhase`. (Although the performance difference is likely not measurable on modern hardware.)
 * What I would like to do longer term is add an obvious "correct" way to do a full-screen pass. We do this in 4 places in the current render graph (see cmd_draw(3, 0)) vs 5 call sites for `write_view_phase`.

However, since I don't think the current implementation is ideal, if you feel strongly that doing it as a feature is better, we can do it that way now. I would like to find a better way to support inserting a "full screen pass" later that is less code than either of these approaches because it seems like a common operation.